crates/sui-core/src/authority.rs

```diff
@@ -3264,16 +3270,39 @@ impl AuthorityState {
                     ?quorum_threshold,
                     ?buffer_stake_bps,
                     ?effective_threshold,
-                    ?next_protocol_version,
+                    ?proposed_protocol_version,
                     ?packages,
                     "support for upgrade"
                 );
 
                 let has_support = total_votes >= effective_threshold;
-                has_support.then_some((next_protocol_version, packages))
+                has_support.then_some((proposed_protocol_version, packages))
             })
-            // if there was no majority, there is no upgrade
-            .unwrap_or((current_protocol_version, vec![]))
+    }
+
+    fn choose_protocol_version_and_system_packages(
+        current_protocol_version: ProtocolVersion,
+        protocol_config: &ProtocolConfig,
+        committee: &Committee,
+        capabilities: Vec<AuthorityCapabilities>,
+        buffer_stake_bps: u64,
+    ) -> (ProtocolVersion, Vec<ObjectRef>) {
+        std::iter::once(Some((current_protocol_version, Vec::new())))
+            .chain((1..).map(|i| {
+                let proposed_next_epoch_protocol_version = current_protocol_version + i;
+                Self::is_protocol_version_supported(
+                    current_protocol_version,
+                    proposed_next_epoch_protocol_version,
+                    protocol_config,
+                    committee,
+                    capabilities.clone(),
+                    buffer_stake_bps,
+                )
+            }))
+            .take_while(|x| x.is_some())
+            .last()
+            .unwrap()
+            .unwrap()
```

Comment:
Nit: This may be neater in a more imperative style:

```rust
let mut next_protocol_version = current_protocol_version;
let mut system_packages = vec![];
while let Some(version, packages) = Self::is_protocol_version_supported(
    current_protocol_version,
    next_protocol_version + 1,
    protocol_config,
    committee,
    capabilities.clone(),
    buffer_stake_bps
) {
    next_protocol_version = version;
    system_packages = packages;
}

(next_protocol_version, system_packages)    
```

Reply:
But why even use rust if we can't write incomprehensible strings of combinators!

(your way is better, will change)